### PR TITLE
Move scavenger-related CLI functionality to new ScavengerDelegate

### PR DESCRIPTION
**Pull request by rwy0717**
This is a first pass. The example and the empty glue version of the
delegate are implemented. The code changes are under the
OMR_GC_SCAVENGER_DELEGATE ifdef, and disabled by default.

Signed-off-by: Robert Young <rwy0717@gmail.com>

### CODE REVIEW

File: example/glue/ScavengerDelegate.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+#include "omr.h"
+#include "omrExampleVM.hpp"
+#include "omrvm.h"
+#include "OMRVMInterface.hpp"
+#include "ParallelGlobalGC.hpp"
+#include "Scavenger.hpp"
+#include "SlotObject.hpp"
+
+void
+MM_ScavengerDelegate::masterSetupForGC(MM_EnvironmentBase *env)
```
Comment (rwy0717) [inline example/glue/ScavengerDelegate.cpp]: The implementation of this file was taken from the example's CollectorLanguageInterfaceImpl, with the following changes:

* wrapped body in ifdef(OMR_GC_SCAVENGER_DELEGATE)
* renamed CLI to MM_ScavengerDelegate
* deleted virtual keyword

File: example/glue/ScavengerDelegate.hpp [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+class MM_CompactScheme;
+class MM_EnvironmentStandard;
+class MM_ForwardedHeader;
+class MM_MarkMap;
+class MM_MemorySubSpaceSemiSpace;
+
+class MM_ScavengerDelegate : public MM_BaseVirtual {
+
+private:
+	MM_CollectorLanguageInterface* _cli;
```
Comment (rwy0717) [inline example/glue/ScavengerDelegate.hpp]: The example delegate is holding a pointer to the cli, which isn't actually used. The pointer is initialized in the constructor. I want to assess if openj9's scavenger will need the CLI before I remove.

File: example/glue/ScavengerDelegate.hpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+#endif /* defined(OMR_ENV_DATA64) && !defined(OMR_GC_COMPRESSED_POINTERS) */
+
+	/**
+	 * This method will be called on the master GC thread after each scavenger cycle, successful or
+	 * otherwise. It may be used for reporting or other tasks as required.
+	 *
+	 * @param[in] env The environment for the calling thread.
+	 * @param[in] scavengeSuccessful Indicates whether the scavenge cycle completed normally or aborted.
+	 */
+	void reportScavengeEnd(MM_EnvironmentBase * env, bool scavengeSuccessful);
```
Comment (rwy0717) [inline example/glue/ScavengerDelegate.hpp]: function names were taken from the CLI, with the `scavenger_` prefix removed.

File: glue/ScavengerDelegate.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
+/*******************************************************************************
```
Comment (rwy0717) [inline glue/ScavengerDelegate.cpp]: The implementation of this file was taken from the empty glue's CLI, with functions and classes renamed.

File: gc/base/standard/Scavenger.cpp [hunk 5]
```diff
@@ -382,7 +382,11 @@ MM_Scavenger::masterSetupForGC(MM_EnvironmentStandard *env)
 	clearCycleGCStats(env);
```
Comment (rwy0717) [inline gc/base/standard/Scavenger.cpp]: In this file, every usage of the CLI is replaced by the corresponding call to the delegate. The code should be functionally equivalent.

File: example/glue/ScavengerDelegate.hpp (hunk 2 above)
Comment (rwy0717) [reply example/glue/ScavengerDelegate.hpp]: Removed the CLI pointer.

**Comment by rwy0717** [comment]
Added an initialize and tearDown member-functions to the delegate, called during scavenger init/teardown. There are no newInstance/kill functions, because the delegate is embedded in the scavenger, not a separate allocation.

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
Nothing here seems too complicated, just a refactoring.  Hopefully we get this in soon and can update the OpenJ9 side of things.

**Comment by youngar** [comment]
@charliegracie Any comments, or can I merge this?

**Comment by rwy0717** [comment]
Regenerated the configure script

**Comment by charliegracie** [comment]
@genie-omr build all

File: example/glue/ScavengerDelegate.cpp (hunk 1 above)
Comment (ktbriggs) [reply example/glue/ScavengerDelegate.cpp]: Can we also remove the remaining `*_` prefixes from the function names? They too seem superfluous at this point...

**Pull request merged**